dockstore-webservice/src/main/java/io/dockstore/webservice/resources/proposedGA4GH/ToolsApiExtendedServiceImpl.java

```diff
@@ -183,42 +169,63 @@ public Response organizationsGet(SecurityContext securityContext) {
     @Override
     public Response toolsIndexGet(SecurityContext securityContext) {
         if (!config.getEsConfiguration().getHostname().isEmpty()) {
-            List<Entry> published = getPublished();
-            try {
-                RestHighLevelClient client = ElasticSearchHelper.restHighLevelClient();
-                // Delete previous indices
-                deleteIndex(client, TOOLS_INDEX);
-                deleteIndex(client, WORKFLOWS_INDEX);
-
-                // Get mapping for tools index
-                URL urlTools = Resources.getResource("queries/mapping_tool.json");
-                String textTools = Resources.toString(urlTools, StandardCharsets.UTF_8);
-
-                // Get mapping for workflows index
-                URL urlWorkflows = Resources.getResource("queries/mapping_workflow.json");
-                String textWorkflows = Resources.toString(urlWorkflows, StandardCharsets.UTF_8);
-
-                // Create indices
-                CreateIndexRequest toolsRequest = new CreateIndexRequest(TOOLS_INDEX);
-                toolsRequest.source(textTools, XContentType.JSON);
-                CreateIndexRequest workflowsRequest = new CreateIndexRequest(WORKFLOWS_INDEX);
-                workflowsRequest.source(textWorkflows, XContentType.JSON);
-                client.indices().create(toolsRequest, RequestOptions.DEFAULT);
-                client.indices().create(workflowsRequest, RequestOptions.DEFAULT);
-
-                // Populate index
-                if (!published.isEmpty()) {
-                    publicStateManager.bulkUpsert(published);
-                }
-            } catch (IOException e) {
-                LOG.error("Could not create elastic search index", e);
-                throw new CustomWebApplicationException("Search indexing failed", HttpStatus.SC_INTERNAL_SERVER_ERROR);
-            }
-            return Response.ok().entity(published.size()).build();
+            clearIndex();
+            int totalProcessed = 0;
+            LOG.info("Starting GA4GH batch processing");
+            totalProcessed += indexDAO(toolDAO);
+            LOG.info("Processed {} tools", totalProcessed);
+            totalProcessed += indexDAO(workflowDAO);
+            LOG.info("Processed {} tools and workflows", totalProcessed);
+            totalProcessed += indexDAO(appToolDAO);
+            LOG.info("Processed {} tools, workflows, and apptools", totalProcessed);
+            return Response.ok().entity(totalProcessed).build();
         }
         return Response.ok().entity(0).build();
     }
 
+    private int indexDAO(EntryDAO entryDAO) {
+        int processed = 0;
+        List<? extends Entry> published;
+        do {
+            published = entryDAO.findAllPublished(processed, ES_BATCH_INSERT_SIZE, null, "gitUrl", "asc");
```

Comment:
I think objects for all published entries will still end up in memory, even though you're now fetching them in batches. Which someday could be an issue, if we get enough content. I wonder if it's worth doing a `clear()` or `evict()`, but I don't know all the Hibernate implications and if that would really help.

Reply:
I'd hope that the session manager would only track dirty objects, but it's worth a try 
https://docs.jboss.org/hibernate/orm/6.1/userguide/html_single/Hibernate_User_Guide.html#_evicting_entities

Reply:
Yeah, I'm pretty sure the current session keeps all of the retrieved objects around.  A `currentSession.clear()`, possibly preceded by a `flush()`, after the `indexBatch` would be the sledgehammer approach.  Or, instead, perhaps safer, you could evict each individual entry one-by-one, preventing any important objects in the code that called this code from becoming detached (if they exist).

Reply:
Think its just `clear` or `evict` , not `flush` We're not making changes 

```
 Whenever you pass an object to save(), update() or saveOrUpdate() and whenever you retrieve an object using load(), get(), list(), iterate() or scroll(), that object is added to the internal cache of the Session.

When flush() is subsequently called, the state of that object will be synchronized with the database. If you do not want this synchronization to occur or if you are processing a huge number of objects and need to manage memory efficiently, the evict() method may be used to remove the object and its collections from the first-level cache.
```

As for whether Hibernate tracks state of unchanged objects, I did find this 
```
Historically Hibernate only supported diff-based dirty calculation for determining which entities in a persistence context have changed. This essentially means that Hibernate would keep track of the last known state of an entity in regards to the database (typically the last read or write). 
```
https://docs.jboss.org/hibernate/orm/6.1/userguide/html_single/Hibernate_User_Guide.html#BytecodeEnhancement-dirty-tracking

In other words, I don't think it needs to store the whole object in memory if nothing changed, but it probably doesn't hurt to explicitly clear/evict.